doc/specs/#674 - PowerShell cmdlets.md

```diff
@@ -0,0 +1,434 @@
+---
+author: Demitrius Nelon @denelon
+created on: 2021-08-06
+last updated: 2021-08-27
+issue id: 674
+---
+
+# When do we begin to see the PowerShell cmdlets?
+
+"For [#674](https://github.com/microsoft/winget-cli/issues/674)"
+
+## Abstract
+
+This specification is to begin drafting PowerShell cmdlets for the Windows Package Manager. @doctordns submitted the issue
+and this specification is intended to capture the intent for the syntax. The PowerShell Team has recently released
+[PowerShell Crescendo Preview.3](https://devblogs.microsoft.com/powershell/announcing-powershell-crescendo-preview-3/). We
+are considering using this framework to help model the syntax and create a module so users can experiment with the syntax.
+Several of the winget.exe commands provide output that is not representative of properly formatted objects, so there will
+likely be gaps to address natively in the COM interface for the Windows Package Manager, or we will provide an argument to output the results from the client to JSON.
+
+## Inspiration
+
+@ThomasNieto was the first to create an Issue for [Native PowerShell Support](https://github.com/microsoft/winget-cli/issues/221).
+This is one of the [most desired features](https://github.com/microsoft/winget-cli/issues?page=1&q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3AIssue-Feature) for the Windows Package Manager.
+
+## Solution Design
+
+The initial solution will start with [PowerShell Crescendo Preview.3](https://devblogs.microsoft.com/powershell/announcing-powershell-crescendo-preview-3/).
+Many of the major considerations are related to proper PowerShell nouns and verbs. @MilkFather suggested [command aliases](https://github.com/microsoft/winget-cli/issues/380) for the winget.exe
+client and @Jaifroid also suggested [find](https://github.com/microsoft/winget-cli/issues/1299) rather than search.
+
+## UI/UX Design
+
+The main UI/UX differences will be in the command names, parameters, and how the output is formatted for users. For the sake of
+consistency the sub headings below will map to existing commands in winget.exe to begin drafting some syntax proposals. In order
+to keep the discussion focussed, I'll create GitHub discussions for each of these and link them here as we work through driving
+consensus.
+
+The PowerShell module name will be "WindowsPackageManager"
+The Adjective "WinGet" may change, but is used consistently in the cmdlets until such time as we agree on the proper term.
+
+>Note: Naming things is hard, and one of the conventions of PowerShell is to make nouns singular.
+---
+
+### Client command `winget install`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Install-WinGetPackage
+```
+**Parameters**
+
+-Manifest
+
+-PackageIdentifier
```

Comment:
Since this parameter will be used quite often I think it would make sense to add a parameter alias:

```c#
[Alias("Id")]
```

Reply:
surely it should be Name?

Reply:
Package ID and Name are two different things. The Name parameter is two lines below

Reply:
Can you have two packages with the same name? Likewise Package-ID?  What is the difference?

Reply:
Take a look at the schema for an explanation of the schema: https://github.com/microsoft/winget-cli/blob/master/schemas/JSON/manifests/v1.0.0/manifest.singleton.1.0.0.json

Both are arbitrary strings, only the ID has to be unique. It's the reason you basically can't install anything with winget unless you use `--id` because it always finds multiple results